parser/parser.go

```diff
@@ -6,127 +6,168 @@ import (
 	"errors"
 	"fmt"
 	"io"
-	"log/slog"
-	"slices"
+	"strings"
 )
 
 type Command struct {
 	Name string
 	Args string
+	bytes.Buffer
 }
 
-func (c *Command) Reset() {
-	c.Name = ""
-	c.Args = ""
-}
-
-func Parse(reader io.Reader) ([]Command, error) {
-	var commands []Command
-	var command, modelCommand Command
+func Parse(r io.Reader) ([]Command, error) {
+	var cmds []Command
+	var cmd Command
+	var b bytes.Buffer
+
+	var quotes int
+
+	s := stateName
+	br := bufio.NewReader(r)
+	for {
+		r, _, err := br.ReadRune()
+		if errors.Is(err, io.EOF) {
+			break
+		} else if err != nil {
+			return nil, err
+		}
 
-	scanner := bufio.NewScanner(reader)
-	scanner.Buffer(make([]byte, 0, bufio.MaxScanTokenSize), bufio.MaxScanTokenSize)
-	scanner.Split(scanModelfile)
-	for scanner.Scan() {
-		line := scanner.Bytes()
+		if _, err := cmd.WriteRune(r); err != nil {
+			return nil, err
+		}
 
-		fields := bytes.SplitN(line, []byte(" "), 2)
-		if len(fields) == 0 || len(fields[0]) == 0 {
+		// trim leading whitespace
+		if (space(r) || newline(r)) && b.Len() == 0 {
 			continue
 		}
 
-		switch string(bytes.ToUpper(fields[0])) {
-		case "FROM":
-			command.Name = "model"
-			command.Args = string(bytes.TrimSpace(fields[1]))
-			// copy command for validation
-			modelCommand = command
-		case "ADAPTER":
-			command.Name = string(bytes.ToLower(fields[0]))
-			command.Args = string(bytes.TrimSpace(fields[1]))
-		case "LICENSE", "TEMPLATE", "SYSTEM", "PROMPT":
-			command.Name = string(bytes.ToLower(fields[0]))
-			command.Args = string(fields[1])
-		case "PARAMETER":
-			fields = bytes.SplitN(fields[1], []byte(" "), 2)
-			if len(fields) < 2 {
-				return nil, fmt.Errorf("missing value for %s", fields)
+		switch s {
+		case stateName, stateParameter:
+			if alpha(r) || number(r) || (r == '_' && s == stateParameter) {
+				if _, err := b.WriteRune(r); err != nil {
+					return nil, err
+				}
+			} else if space(r) {
+				cmd.Name = strings.ToLower(b.String())
+				b.Reset()
+
+				if cmd.Name == "from" {
+					cmd.Name = "model"
+				}
+
+				switch cmd.Name {
+				case "parameter":
+					s = stateParameter
+				case "message":
+					s = stateMessage
+				default:
+					s = stateArgs
+				}
+			} else if newline(r) {
+				return nil, fmt.Errorf("missing value for [%s]", b.String())
+			} else {
+				return nil, fmt.Errorf("unexpected rune %q for state %d", r, s)
```

Comment:
This is going to trip up on comments which start with "# this is a comment". I think you'll need another state like `stateComment` which starts when you find a "#" at the beginning of the line and then ends when you get a newline.

Reply:
I missed it since there wasn't any tests. should be fixed now